independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java

```diff
@@ -221,22 +240,57 @@ public ClientImpl build() {
             options.setVerifyHost(false);
         }
 
-        char[] effectiveTrustStorePassword = trustStorePassword == null ? EMPTY_CHAR_ARARAY : trustStorePassword;
-        Buffer keyStore = asBuffer(this.keyStore, keystorePassword);
-        Buffer trustStore = asBuffer(this.trustStore, effectiveTrustStorePassword);
-        if (keyStore != null || trustStore != null) {
-            options = options.setSsl(true);
-            if (keyStore != null) {
-                JksOptions jks = new JksOptions();
-                jks.setValue(keyStore);
-                jks.setPassword(new String(keystorePassword));
-                options = options.setKeyStoreOptions(jks);
+        if (certificate != null && key != null) {
+            try {
+                byte[] certBytes = getFileContent(certificate);
+                byte[] keyBytes = getFileContent(key);
+
+                PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions()
+                        .setCertValue(Buffer.buffer(certBytes))
+                        .setKeyValue(Buffer.buffer(keyBytes));
+                options.setKeyCertOptions(pemKeyCertOptions);
```

Comment:
Hm... I just followed what we are already doing for gRPC client

Reply:
maybe it's also missing for gRPC :-D

Reply:
Fair enough :)

Reply:
In that case we would need the same part of keys / certificates, correct? That also means making the names of the properties explicit.

Reply:
Yes, in the HTTP server we use the following:

```java
@ConfigGroup
public class CertificateConfig {


    /**
     * The list of path to server certificates using the PEM format.
     * Specifying multiple files requires SNI to be enabled.
     */
    @ConfigItem
    public Optional<List<Path>> files;

    /**
     * The list of path to server certificates private key files using the PEM format.
     * Specifying multiple files requires SNI to be enabled.
     * <p>
     * The order of the key files must match the order of the certificates.
     */
    @ConfigItem
    public Optional<List<Path>> keyFiles;

    /**
     * An optional keystore that holds the certificate information instead of specifying separate files.
     */
    @ConfigItem
    public Optional<Path> keyStoreFile;

    /**
     * An optional parameter to specify the type of the keystore file.
     * If not given, the type is automatically detected based on the file name.
     */
    @ConfigItem
    public Optional<String> keyStoreFileType;

    /**
     * An optional parameter to specify a provider of the keystore file.
     * If not given, the provider is automatically detected based on the keystore file type.
     */
    @ConfigItem
    public Optional<String> keyStoreProvider;

    /**
     * A parameter to specify the password of the keystore file.
     * If not given, and if it can not be retrieved from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}
     */
    @ConfigItem(defaultValueDocumentation = "password")
    public Optional<String> keyStorePassword;

    /**
     * A parameter to specify a {@linkplain CredentialsProvider} property key,
     * which can be used to get the password of the key
     * store file from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}
     */
    @ConfigItem
    public Optional<String> keyStorePasswordKey;

    /**
     * An optional parameter to select a specific key in the keystore.
     * When SNI is disabled, and the keystore contains multiple
     * keys and no alias is specified; the behavior is undefined.
     */
    @ConfigItem
    public Optional<String> keyStoreKeyAlias;

    /**
     * An optional parameter to define the password for the key,
     * in case it is different from {@link #keyStorePassword}
     * If not given, it might be retrieved from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}.
     */
    @ConfigItem
    public Optional<String> keyStoreKeyPassword;

    /**
     * A parameter to specify a {@linkplain CredentialsProvider} property key,
     * which can be used to get the password for the key from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}
     */
    @ConfigItem
    public Optional<String> keyStoreKeyPasswordKey;

    /**
     * An optional trust store that holds the certificate information of the trusted certificates.
     */
    @ConfigItem
    public Optional<Path> trustStoreFile;

    /**
     * An optional parameter to specify the type of the trust store file.
     * If not given, the type is automatically detected based on the file name.
     */
    @ConfigItem
    public Optional<String> trustStoreFileType;

    /**
     * An optional parameter to specify a provider of the trust store file.
     * If not given, the provider is automatically detected based on the trust store file type.
     */
    @ConfigItem
    public Optional<String> trustStoreProvider;

    /**
     * A parameter to specify the password of the trust store file.
     * If not given, it might be retrieved from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}.
     */
    @ConfigItem
    public Optional<String> trustStorePassword;

    /**
     * A parameter to specify a {@linkplain CredentialsProvider} property key,
     * which can be used to get the password of the trust store file from {@linkplain CredentialsProvider}.
     *
     * @see {@link #credentialsProvider}
     */
    @ConfigItem
    public Optional<String> trustStorePasswordKey;

    /**
     * An optional parameter to trust a single certificate from the trust store rather than trusting all certificates in the
     * store.
     */
    @ConfigItem
    public Optional<String> trustStoreCertAlias;

}

```

Because of mTLS, we need more or less the same set.

Reply:
Wow, that's gonna take some refactoring then...

Reply:
so, there are 2 aspects:

- trust store
- key/cert

On the client side, the trust store is the most essential part, but key/cert is used when using mTLS.

The trust store (basically a list of certs) can be a JKS file (with an alias and password - a single file), a p12 file (with alias and password), or a pem file (list of files, one cert per file)

The key/cert can be:

- Pem files for the key and the certs - for both, it may be a list, even if in the last majority of the case, it will be one key (the private key) and one cert (the cert)
- a JKS file or a P12 file with an alias and password

When using pem file, the extension can be `.pem`, but also `.key` for the private key, `.crt` for the certs and trust store.

Reply:
For the gRPC client, I will likely use the following:

```
package io.quarkus.grpc.runtime.config;

import io.quarkus.runtime.annotations.ConfigGroup;
import io.quarkus.runtime.annotations.ConfigItem;

import java.nio.file.Path;
import java.util.List;
import java.util.Optional;

/**
 * Shared configuration for setting up client-side TLS/SSL.
 */
@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
@ConfigGroup
public class SslClientConfig {
    /**
     * The classpath path or file path to a server certificate or certificate chain in PEM format.
     * This parameter is only used when mutual TLS is enabled.
     *
     * @deprecated Use 'certificates' instead
     */
    @ConfigItem
    @Deprecated
    public Optional<Path> certificate;

    /**
     * The classpath paths or file paths to <strong>server</strong> certificates or certificates chain in PEM format.
     * This parameter is only used when mutual TLS is enabled.
     */
    @ConfigItem
    public Optional<List<Path>> certificates;

    /**
     * The classpath path or file path to the corresponding certificate private key file in PEM format.
     * This parameter is only used when mutual TLS is enabled.
     *
     * @deprecated Use 'keys' instead
     */
    @ConfigItem
    @Deprecated
    public Optional<Path> key;

    /**
     * The classpath path or file path to the corresponding certificate private key file in PEM format.
     * This parameter is only used when mutual TLS is enabled.
     * <p>
     * When multiple paths are specified the order must match the order of the certificates.
     */
    @ConfigItem
    public Optional<List<Path>> keys;

    /**
     * An optional trust store which holds the certificate information of the certificates to trust.
     * <p>
     * The classpath path or file path to a JKS or P12 key store. For backward compatibility reasons, it can also be a PEM file.
     * However, in this case, {@link #trustedCertificates} should be used instead.
     * <p>
     * For P12 and JKS, the truststore may contain multiple certificates.
     * If the alias is set, only the one certificate and key pair will be used (matching the alias).
     * Otherwise, SNI will be used (SNI will need to be enabled).
     */
    @ConfigItem
    public Optional<Path> trustStore;

    /**
     * Optional list of paths which hold the information of the certificates to trust in PEM format.
     * <p>
     * The trust store can be either on classpath or in an external file.
     */
    @ConfigItem
    public Optional<List<Path>> trustedCertificates;


    /**
     * The classpath path or file path to a JKS or P12 key store.
     * <p>
     * The key store may contain multiple certificates and keys.
     * If the alias is set, only the one certificate and key pair will be used (matching the alias).
     * Otherwise, SNI will be used (SNI will need to be enabled).
     */
    @ConfigItem
    public Optional<Path> keyStore;

    /**
     * An optional parameter to specify a provider of the keystore file.
     * If not given, the provider is automatically detected based on the keystore file type.
     */
    @ConfigItem
    public Optional<String> keyStoreProvider;

    /**
     * A parameter to specify the password of the keystore file.
     */
    @ConfigItem(defaultValueDocumentation = "password")
    public Optional<String> keyStorePassword;

    /**
     * An optional parameter to select a specific key in the keystore.
     * When SNI is disabled, and the keystore contains multiple
     * keys and no alias is specified; the behavior is undefined.
     */
    @ConfigItem
    public Optional<String> keyStoreAlias;


    /**
     * An optional parameter to specify a provider of the trust store file.
     * If not given, the provider is automatically detected based on the trust store file type.
     */
    @ConfigItem
    public Optional<String> trustStoreProvider;

    /**
     * A parameter to specify the password of the trust store file.
     * <p>
     * Passwords are only supported for PKCS12 and JKS key stores
     */
    @ConfigItem(defaultValueDocumentation = "password")
    public Optional<String> trustStorePassword;


    /**
     * An optional parameter to trust a single certificate from the trust store rather than trusting all certificates in the
     * store.
     * <p>
     * Alias is only supported for PKCS and JKS trust stores.
     */
    @ConfigItem
    public Optional<String> trustStoreAlias;

}

```

Reply:
Hopefully next week I'll find th e courage to update this PR with that change :)

Reply:
Hold on until further instructions :-D 

I'm doing an inventory of how to configure TLS in our clients... We have enough approaches to open a museum.

Reply:
> We have enough approaches to open a museum

Hahaha

> Hold on until further instructions

Will do :)